dask/dataframe/tests/test_multi.py

```diff
@@ -581,26 +580,36 @@ def test_concat(join):
         assert_eq(result, expected)
 
 
-@pytest.mark.parametrize("join", ["inner", "outer"])
-def test_concat_different_dtypes(join):
+values = [1, 1.0, "one", pd.to_datetime("1970-01-01")]
+
+
+@pytest.mark.parametrize(
+    "value_pair",
+    [
+        (values[i], values[j])
+        for i in range(len(values))
+        for j in range(len(values))
+        if i < j
+    ],
+)
+def test_concat_different_dtypes(value_pair):
     # check that the resulting dataframe has coherent dtypes
     # refer to https://github.com/dask/dask/issues/4685
-    pdf1 = pd.DataFrame(
-        {"x": [1, 2, 3, 4, 6, 7], "y": list("abcdef")}, index=[1, 2, 3, 4, 6, 7]
-    )
-    ddf1 = dd.from_pandas(pdf1, 2)
-    pdf2 = pd.DataFrame(
-        {"x": [1.0, 2.0, 3.0, 4.0, 6.0, 7.0], "y": list("abcdef")},
-        index=[8, 9, 10, 11, 12, 13],
-    )
-    ddf2 = dd.from_pandas(pdf2, 2)
+    df_1 = pd.DataFrame({"x": [value_pair[0]]})
+    df_2 = pd.DataFrame({"x": [value_pair[1]]})
+    df = pd.concat([df_1, df_2], axis=0)
+
+    pandas_dtype = df["x"].dtype
 
-    expected = pd.concat([pdf1, pdf2], join=join)
-    result = dd.concat([ddf1, ddf2], join=join)
-    assert_eq(expected, result)
+    ddf_1 = dd.from_pandas(df_1, npartitions=1)
+    ddf_2 = dd.from_pandas(df_2, npartitions=1)
+    ddf = dd.concat([ddf_1, ddf_2], axis=0)
 
-    dtypes_list = dask.compute(*[part.dtypes for part in result.to_delayed()])
-    assert len(set(map(str, dtypes_list))) == 1  # all the same
+    dask_dtypes = ddf.map_partitions(lambda x: x.dtypes).compute().to_list()
+
+    assert (
+        dask_dtypes == ["datetime64[ns]", "datetime64[ns]"] and pandas_dtype == "object"
+    ) or dask_dtypes == [pandas_dtype, pandas_dtype]
```

Comment:
Why isn't this just `assert all(dask_dtypes == pandas_dtype)`? Are there cases where we don't match the pandas result?

And if you're interested, a "dtype consistency" check in `assert_eq` may be useful. An assert that all the partitions have the same dtypes. That should probably wait for a different PR in case it turns up other issues.

Reply:
As mentioned above, we don't match the pandas dtype for when concatenating dataframes with `datetime` and `float` columns. I thought `assert dask_dtypes == [pandas_dtype, pandas_dtype]` was more explicit, but `assert all(dask_dtypes == pandas_dtype)` should work just as well.

Reply:
Sorry, I missed that comment earlier. Float and datetime should definitely result in object dtype. What do we get? Do you know why we don't match pandas?

Reply:
We get datetime. The `methods.concat()` in this line
```meta = methods.concat([df._meta for df in dfs], join=join, filter_warning=False)```
returns a dataframe with dtype datetime.

Reply:
Are you able to look into why we're getting a different result?

Reply:
The issue seems to lie with pandas:
```python
import pandas as df

df_1 = pd.DataFrame({"x": [0.0]})
df_2 = pd.DataFrame({"x": [pd.to_datetime("1970-01-01")]})

assert pd.concat([df_1, df_2])["x"].dtype == "object"
assert pd.concat([df_1.iloc[0:0], df_2.iloc[0:0]])["x"].dtype == "datetime64[ns]"
```

Since `_meta` are empty dataframes, we get the latter result.

Reply:
Thanks. I opened https://github.com/pandas-dev/pandas/issues/32934 to track that.

So we can either work around it in the code, or segment that test. I'm inclined to not work around it in the code.

Are you able to write the correct assert, but xfail the test for the datetime & float case?

Reply:
Sure.